File: python/c07_6_counting_tilings.py
import unittest
from math import pi,cos,prod
from itertools import product

def solve(n,m):
    def isvalid(prev,curr):
        for j in range(m):
            if prev[j]=='A':
                if curr[j]!='V' or j>0 and curr[j-1]=='<':
                    return False
            elif j>0 and curr[j-1]=='<':
                if curr[j]!='>':
                    return False
            elif curr[j]!='A' and not (j<m-1 and curr[j]=='<'):
                return False
        return True
    if n<m:
        n,m=m,n
    dp={'V'*m:1}
    for _ in range(n):
        ndp={}
        for curr in product('V<A>',repeat=m):
            sum_=sum(v for prev,v in dp.items() if isvalid(prev,curr))
            if sum_>0:
                ndp[curr]=sum_
        dp=ndp
    return sum(v for prev,v in dp.items() if 'A' not in prev)

def solve2(n,m):
    def isvalid(prev,curr):
        open_=False
        for j in range(m):
            if prev[j]=='A':
                if curr[j]!='O' or open_:
                    return False
            elif open_:
                if curr[j]!='O':
                    return False
                open_=False
            elif curr[j]=='O':
                if j>=m-1:
                    return False
                open_=True
        return True
    if n<m:
        n,m=m,n
    dp={'O'*m:1}
    for _ in range(n):
        ndp={}
        for curr in product('OA',repeat=m):
            sum_=sum(v for prev,v in dp.items() if isvalid(prev,curr))
            if sum_>0:
                ndp[curr]=sum_
        dp=ndp
    return sum(v for prev,v in dp.items() if 'A' not in prev)

def solve3(n,m):
    def cosn(n):
        return [cos(pi*a/(n+1))**2 for a in range(1,(n+3)//2)]
    cn,cm=cosn(n),cosn(m)
    return round(prod(prod(4*(k+l) for l in cm) for k in cn))

class Test(unittest.TestCase):
    def test(self):
        for s in (solve,solve2,solve3):
            with self.subTest(s=s):
                self.assertEqual(s(0,0),1)
                self.assertEqual(s(4,7),781)

if __name__=='__main__':
    unittest.main()

File: c++/c07_6_counting_tilings.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;
#include "util.h"

namespace {
	ll solve1(ll n, ll m) {
		if (n < m)
			swap(n, m);
		auto isvalid = [m](const string& prev, const string& curr) {
			for (ll j = 0; j < m; ++j) {
				if (prev[j] == 'A') {
					if (curr[j] != 'V' || j > 0 && curr[j - 1] == '<') {
						return false;
					}
				}
				else if (j > 0 && curr[j - 1] == '<') {
					if (curr[j] != '>') {
						return false;
					}
				}
				else if (curr[j] != 'A' && !(j < m - 1 && curr[j] == '<'))
					return false;
			}
			return true;
			};
		unordered_map<string, ll> dp;
		dp[string(m, 'V')] = 1;
		for (ll i = 0; i < n; ++i) {
			unordered_map<string, ll> ndp;
			vector<ll> vec(m);
			do {
				auto curr = vec | views::transform([](ll i) { return "V<A>"[i]; }) | ranges::to<string>();
				ll sum = 0;
				for (const auto& [prev, v] : dp) {
					if (isvalid(prev, curr)) {
						sum += v;
					}
				}
				if (sum > 0)
					ndp[curr] = sum;
			} while (next_product(vec, 4));
			dp = move(ndp);
		}
		ll ans = 0;
		for (const auto& [prev, v] : dp) {
			if (!prev.contains('A')) {
				ans += v;
			}
		}
		return ans;
	}

	ll solve2(ll n, ll m) {
		if (n < m)
			swap(n, m);
		auto isvalid = [m](const string& prev, const string& curr) {
			bool open = false;
			for (ll j = 0; j < m; ++j) {
				if (prev[j] == 'A') {
					if (curr[j] != 'O' || open) {
						return false;
					}
				}
				else if (open) {
					if (curr[j] != 'O')
						return false;
					open = false;
				}
				else if (curr[j] == 'O') {
					if (j >= m - 1)
						return false;
					open = true;
				}
			}
			return true;
			};
		unordered_map<string, ll> dp;
		dp[string(m, 'O')] = 1;
		for (ll i = 0; i < n; ++i) {
			unordered_map<string, ll> ndp;
			vector<ll> vec(m);
			do {
				auto curr = vec | views::transform([](ll i) { return "OA"[i]; }) | ranges::to<string>();
				ll sum = 0;
				for (const auto& [prev, v] : dp) {
					if (isvalid(prev, curr)) {
						sum += v;
					}
				}
				if (sum > 0)
					ndp[curr] = sum;
			} while (next_product(vec, 2));
			dp = move(ndp);
		}
		ll ans = 0;
		for (const auto& [prev, v] : dp) {
			if (!prev.contains('A')) {
				ans += v;
			}
		}
		return ans;
	}

	ll solve3(ll n, ll m) {
		double ret = 1;
		for (ll a = 1; a <= ((n + 1) / 2); ++a) {
			for (ll b = 1; b <= ((m + 1) / 2); ++b) {
				double ca = cos(numbers::pi * a / (n + 1)), cb = cos(numbers::pi * b / (m + 1));
				ret *= 4 * (ca * ca + cb * cb);
			}
		}
		return ll(round(ret));
	}

	void test(auto f) {
		EXPECT_EQ(f(0, 0), 1);
		EXPECT_EQ(f(4, 7), 781);
	}
}

TEST(C076CountingTilings, solve1) {
	test(solve1);
}

TEST(C076CountingTilings, solve2) {
	test(solve2);
}

TEST(C076CountingTilings, solve3) {
	test(solve3);
}
